### ISSUE TITLE
JUnit 5 support

### ISSUE THREAD

**Issue by kthoms** [eclipse/xtext-core#666]
Xtext's testing framework should support writing test for JUnit 5.

Most notably is that the `@RunWith` annotation is not available anymore and [has been replaced by the more flexible `@ExtendWith` annotation](http://www.baeldung.com/junit-5-runwith). However, we still need the logic that is employed by `XtextRunner`, and preserve `@InjectWith`.

This basically means that `XtextRunner`'s logic has to be transferred to the new JUnit 5 extension API. This concept should be added to `org.eclipse.xtext.testing`, introducing new dependencies on the JUnit 5 framework.

Clients may use either JUnit 4 or 5, or for transition both. Since JUnit 5 API is providing different packages it is possible to manage both.

As a result clients should be able to leverage JUnit 5 to write and execute Xtext based tests with typical dependency injection.

**Comment by kthoms** [comment]
CQ for JUnit 5.0 usage, approved: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=15984

**Comment by kthoms** [comment]
I think we are done. Documentation, of course, must be added.

**Comment by kthoms** [comment]
The test projects do not need direct dependencies on the JUnit engine bundles. Tycho is smart enough to use these dependencies at runtime. It even harms, see https://bugs.eclipse.org/bugs/show_bug.cgi?id=534587 for that.

Instead, the test projects only need dependencies on the JUnit API. When using the JUnit library container in the project configuration the required bundles are automatically inserted, including the engine bundles for JU5.

JU4
<img width="793" alt="screenshot 101" src="https://user-images.githubusercontent.com/265597/40128066-0c2e40fc-5931-11e8-83d2-da5414837470.png">

JU5
<img width="786" alt="screenshot 103" src="https://user-images.githubusercontent.com/265597/40128072-0ed8a874-5931-11e8-832b-6ace358cd3ef.png">

**Comment by cdietrich** [comment]
Junit 5 tests are not executed in pure maven projects

**Comment by cdietrich** [comment]
Junit5 test are NOT executed in pure gradle projects

**Comment by cdietrich** [comment]
for gradle the following seems to work

```
dependencies {
	compile project(':org.xtext.example.mydsl1')
	testCompile "org.junit.jupiter:junit-jupiter-api:5.1.0"
	testCompile "org.junit.jupiter:junit-jupiter-params:5.2.0"
	testCompile "org.eclipse.xtext:org.eclipse.xtext.testing:${xtextVersion}"
	testCompile "org.eclipse.xtext:org.eclipse.xtext.xbase.testing:${xtextVersion}"
	testRuntime "org.junit.jupiter:junit-jupiter-engine:5.1.0"
}

test {
	useJUnitPlatform()

	testLogging {
		events "passed", "skipped", "failed"
	}

	reports {
		html.enabled = true
	}
}

```

**Comment by kthoms** [comment]
maven is similar, PR will come

**Comment by cdietrich** [comment]
@kthoms maven fix does not seem to work in maven layout (code goes to runtime)

**Comment by kthoms** [comment]
That problem already exists on master. Therefore different issue?

**Comment by cdietrich** [comment]
i dont get. with junit 4 no issue

**Comment by cdietrich** [comment]
there is something like
``«IF testProject.isInlined»``` in runtimedescriotor

**Comment by kthoms** [comment]
IMHO it is correct to have the tests in a plain Maven project with Maven source layout to be in `src/test/java`.
But the surefire config has to be in the runtime project's pom.

**Comment by cdietrich** [comment]
yes but exactly that is missing for your maven and gradle fixes

**Comment by kthoms** [comment]
did not consider that, sorry. good catch.

**Comment by cdietrich** [comment]
yes generate surefire and gradle test customization only for plain layout

**Comment by cdietrich** [comment]
if you do prs for that will review when awake again

**Comment by kthoms** [comment]
for Maven I'd better put this into the pluginManagement of the parent POM. Then it should work for bith cases.

**Comment by cdietrich** [comment]
retested pure maven stuff. looks good

**Comment by cdietrich** [comment]
i wonder if we should enhance the classpathcontainer thingy to support

```
	<classpathentry kind="con" path="org.eclipse.jdt.junit.JUNIT_CONTAINER/5">
		<attributes>
			<attribute name="test" value="true"/>
		</attributes>
	</classpathentry>
```

unfortunately the support for that only works if we fix it for the source folders too.
=> should we do that with a separate issue in 2.15?

**Comment by kthoms** [comment]
I already created an issue for test source folders:
https://github.com/eclipse/xtext-eclipse/issues/677

Might be worth to mention that in the issue. IMHO too late for 2.14 since we have to check that on Oxygen also and could cause trouble.

**Comment by kthoms** [comment]
For the Inlined Maven case the JUnit containers are not configured. It is expected that there would be a "Junit 5" entry in the package explorer.

![screenshot 113](https://user-images.githubusercontent.com/265597/40166807-c37bcb4c-59bf-11e8-8dec-40fdcbf31594.png)

**Comment by cdietrich** [comment]
see https://github.com/eclipse/xtext-core/issues/666#issuecomment-389749054

**Comment by cdietrich** [comment]
@kthoms anything left that we want to fix with 2.14?

**Comment by kthoms** [comment]
With fixing of #741 I close this for 2.14 now.
